Pypi hash retrieval

techalchemy opened the pull request:
This is the final piece in fixing post release resolution.  Kenneth had the first piece in 2ed0eb33 and this piece handles resolving hashes against pypi 

* Note that we try both getting all options from PyPI as well as resolving hashes for specific, known packages using the pip resolver due to edge cases where, for example, there is an implicit post release but no base release.  PyPI does not handle these very gracefully on the API side, where they display information and hashes for the RELEASE, but if you try to pip install only the post-release versioning will work.  I.e. system pip only is aware of the post-release version and the API is only aware of the release version information, but the resolver seems able to translate

See #314 #400 #462

nateprewitt commented on pipenv/utils.py:
```diff
@@ -346,3 +364,4 @@ def find_requirements(max_depth=3):
                     if os.path.isfile(r):
                         return r
         raise RuntimeError('No requirements.txt found!')
+
```
Actually let's remove this extra space too. Thanks!

techalchemy commented:
@nateprewitt I fixed those issues + one additional issue with asking pypi for hashes of editable packages which i've just added an exclusion rule on for the moment, not sure if there is a better way to handle that

nateprewitt commented:
Hmm, looks like tests are failing for a variety of reasons. Not sure if this is our CI environment or the actual change being made.

techalchemy commented:
They're failing on master as well, with the same error

kennethreitz commented:
Working on it :)

kennethreitz commented:
@techalchemy i'm confused that you didn't remove any of the other hashing code?

kennethreitz commented:
Ah, I see. Hmm. Does this not just replace my code entirely?

kennethreitz commented:
reverting this — it's downloading files.

kennethreitz commented:
got this merged, with some major simplification :)

techalchemy commented:
**Note that @kennethreitz has a cleverer fix than this PR which introduced substantial lag especially for European users of pipenv -- see 175882048 **

@kennethreitz I didn't remove your hashing code based on a brief point in our discussion earlier.  Namely:

* Your code pulls all hashes for the available versions on PyPI to address potential disparities across different wheels for different python versions
  - This works only if the release is known and resolves correctly via REST API call

* My code traverses the dependency tree for the current setup and uses the pip resolver (via pip-tools) to resolve release/versions and acquire hashes, which addresses
  - This addresses the case where the resolved release and API release do not align
  - This has the appearance of being a drop-in replacement for the previous approach but I am not convinced that it would resolve the issues you mentioned

Pip resolves ibm-db-sa-py3 to 0.3.1.post1 and doesn't resolve 0.3.1, but pypi only returns information for 0.3.1:

```python
>>> from piptools.resolver import Resolver
... from piptools.scripts.compile import get_pip_command
... from piptools.repositories.pypi import PyPIRepository
... constraints = []
... req = "ibm-db-sa-py3>=0.3.1-1"
... import pip
... c = pip.req.InstallRequirement(req=req, comes_from='nowhere')
... constraints.append(c)
... pip_command = get_pip_command()
... pip_args = []
... pip_options, _ = pip_command.parse_args(pip_args)
... import requests
... pypi = PyPIRepository(pip_options=pip_options, session=requests.Session())
... r = Resolver(constraints=constraints, repository=pypi)
... deps = [d for d in r.resolve()]

>>> deps
[<InstallRequirement object: ibm-db-sa-py3==0.3.1.post1 editable=False>, <InstallRequirement object: sqlalchemy==1.1.14 editable=False>]
```

At the time of this writing, the pypi endpoint for my original case is [here](https://pypi.org/pypi/ibm-db-sa-py3/json) and shows (in abbreviated format):

```json
{
	"info": {
		"author": "IBM Application Development Team",
		"author_email": "opendev@us.ibm.com",
		"license": "Apache License 2.0",
		"name": "ibm-db-sa-py3",
		"project_url": "https://pypi.org/project/ibm-db-sa-py3/",
		"release_url": "https://pypi.org/project/ibm-db-sa-py3/0.3.1-1/",
		"requires_dist": [],
		"requires_python": null,
		"summary": "SQLAlchemy support for IBM Data Servers",
		"version": "0.3.1-1"
	},
	"releases": {
		"0.3.0": [{
			"comment_text": "",
			"digests": {
				"md5": "1ec10ad9922f59d0a4223c49829f9652",
				"sha256": "054e05ee6615979b8025a9b6122aa889e1abece0ccdeaa7958f0670ad94ecfe5"
			},
			"filename": "ibm-db-sa-py3-0.3.0.tar.gz",
			"has_sig": false,
			"md5_digest": "1ec10ad9922f59d0a4223c49829f9652",
			"packagetype": "sdist"
		}],
		"0.3.0-1": [{
			"digests": {
				"md5": "ead323db4de46726422e5b6dccc5163e",
				"sha256": "b3bb3ce09ad1fad7460a37b137418a2e990fca006b1505fe313cb98b8893f57d"
			},
			"downloads": 4921,
			"filename": "ibm-db-sa-py3-0.3.0-1.tar.gz",
			"has_sig": false,
			"md5_digest": "ead323db4de46726422e5b6dccc5163e",
			"packagetype": "sdist",
		}],
		"0.3.1": [{
			"digests": {
				"md5": "b5326cdbd61c0dbab2003f62debc8bcb",
				"sha256": "c7d50cc3a8ef9e393c35af01d478a05438159a8c2304bf871e3d5eeb43fb97b2"
			},
			"downloads": 1879,
			"filename": "ibm-db-sa-py3-0.3.1-1.tar.gz",
			"has_sig": false,
			"md5_digest": "b5326cdbd61c0dbab2003f62debc8bcb",
			"packagetype": "sdist",
		}],
		"0.3.1-1": []
	},
	"urls": []
}
```

kennethreitz commented:
✨🍰✨
